[IMP] mail,im_livechat: Modernization of the chat windows

alexkuhn opened the pull request:
Redesign of chat window management for internal users.

Folded conversations are now displayed as ordered bubbles. The
footprint is drastically reduced and browsing active chats is easier.

https://github.com/odoo/enterprise/pull/65211

<img width="2560" alt="Screenshot 2024-06-26 at 12 11 27 (3)" src="https://github.com/odoo/odoo/assets/6569390/c73844fd-3dc5-4f4b-9d4e-36daaed0496a">

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/161922).

alexkuhn commented:
Rebased version of https://github.com/odoo/odoo/pull/136397

alexkuhn commented on addons/im_livechat/static/src/embed/common/livechat_button.js:
```diff
@@ -71,7 +50,8 @@ export class LivechatButton extends Component {
             this.livechatService.available &&
-            this.livechatService.state === SESSION_STATE.NONE
+            this.livechatService.state === SESSION_STATE.NONE &&
+            Object.keys(this.store.ChatWindow.records).length === 0
```
Not crucial, but if we don't place feedback chat window exactly above the button, visitor could have 2 chat windows at the same time.

I made livechat-specific code that keeps chat window exactly above the button, but without it there could be offset. So this is more defensive than anything

phenix-factory commented on addons/mail/static/src/core/common/chat_bubble.scss:
```diff
@@ -0,0 +1,76 @@
+            visibility: visible;
+        }
+        .o-mail-ChatBubble-counter {
+            visibility: hidden;
```
`display: none` should be used as much as possible, because `visibility` will hide the element to a screen reader. 
(see https://developer.mozilla.org/fr/docs/Web/CSS/visibility#accessibilit%C3%A9)

zel-odoo commented:
![1719998161709](https://github.com/odoo/odoo/assets/26395662/fc17e2e5-8cbe-426d-9f1a-6b1dc8d259a5)
The status icon is barely readable, maybe adjust it? And needs more spacing for the right part I guess.

phenix-factory commented on addons/mail/static/src/core/common/chat_window.xml:
```diff
@@ -2,14 +2,13 @@
 
 <t t-name="mail.ChatWindow">
-    <div class="o-mail-ChatWindow"
+    <div class="o-mail-ChatWindow position-fixed bottom-0 overflow-hidden d-flex flex-column"
```
Maybe `.fixed-bottom`?
```suggestion
    <div class="o-mail-ChatWindow fixed-bottom overflow-hidden d-flex flex-column"
```

zel-odoo commented on addons/mail/static/src/core/common/chat_bubble.scss:
```diff
@@ -0,0 +1,76 @@
+
+.o-mail-ChatBubble {
+    height: auto !important;
```
overall, now the js and design team is gradually making everything bootstrap class defined. So if we have the chance to do so, we should follow the trend as well? So that's easier to maintain the code.

alexkuhn replied:
I understand to prefer using bootstrap classes to avoid writing any SCSS, but sometimes it's stupid to write the same classnames in all occurrences of multiple targeted elements.

There are 3 UI elements that have `o-mail-ChatBubble`. One is an actual chat bubble, the 2 others want to look the same, hence I've just reused the classname. It's worse maintenance to ensure they have the same multiple bootstrap classes when you can just share it automatically by defining in SCSS once.

phenix-factory commented on addons/im_livechat/static/tests/embed/chat_window.test.js:
```diff
@@ -44,7 +42,8 @@ test("do not save fold state of temporary live chats", async () => {
     await assertSteps(["fold - folded"]);
-    await click("[title='Close Chat Window']");
+    await contains(".o-mail-ChatBubble");
+    await click("button[title*='Close Chat Window']");
```
Title look precise enough, you don't need the `button`

alexkuhn commented:
> The status icon is barely readable, maybe adjust it? And needs more spacing for the right part I guess.

I added 5px more to the right. For the im status icon, I have moved it slightly more to bottom-right and added a `filter: brightness(1.25)`, which improves the visibility.

The implementation technique we use to make a inner circle matching background around the im status doesn't work here.
Some workaround solutions could be to make a SVG mask on image, so that it actually removes the content on the image.
This isn't trivial and can be hard to make it size-aware with CSS spacings... So better not do it now and consider as a potential future improvement

alexkuhn commented:
@robodoo r+
